### PR TITLE
Add support for Gather/scatter on aarch64, reduce cols idx max on aarch64, and fix multiple issues in eltwise and equations

### PR DESCRIPTION
**Pull request by egeor**
Add support for:

-  Gather/scatter on aarch64
-  Reduce cols idx max on aarch64

It also fix multiple issues in GEMM, eltwise and equations for aarch64. Now the CI is green for gvt3.

### CODE REVIEW

File: samples/eltwise/kernel_test/unary_reduce.tpl [hunk 1]
```diff
@@ -38,7 +40,7 @@ do
   LDO=`echo ${i} | awk -F"_" '{print $4}'`
   echo ${M} ${N} ${LDI} ${LDI}
   N_ADJ=$((${N} + ${N_IDX}))
-  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} ${PREC} ${N_IDX} 0 0 0 ${USE_ACC}
+  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} ${PREC} ${N_IDX} 0 ${IDX_TYPE} ${RECORD_IDX} ${USE_ACC}
```
Comment (alheinecke) [inline samples/eltwise/kernel_test/unary_reduce.tpl:43]: There is still a fix "0" should we also make this variable or remove it from cmd?

File: src/generator_common_aarch64.h [hunk 2]
```diff
@@ -16,6 +16,18 @@
 #include "generator_common.h"
 #include "generator_aarch64_instructions.h"
 
+LIBXSMM_API_INTERN
+void libxsmm_generator_gather_scatter_vreg_asimd_aarch64( libxsmm_generated_code* io_generated_code,
+                                                       const unsigned int      i_is_gather,
```
Comment (alheinecke) [inline src/generator_common_aarch64.h:21]: nit: idention

File: src/generator_common_aarch64.c [hunk 3]
```diff
@@ -158,6 +158,48 @@ void libxsmm_generator_vloadstore_masked_vreg_aarch64_asimd( libxsmm_generated_c
                                                     i_masked_elems == 0 ? 0 : 1 );
 }
 
+LIBXSMM_API_INTERN
+void libxsmm_generator_gather_scatter_vreg_asimd_aarch64( libxsmm_generated_code* io_generated_code,
+                                                       const unsigned int      i_is_gather,
```
Comment (alheinecke) [inline src/generator_common_aarch64.c:163]: nit idention

File: src/generator_common_aarch64.h (hunk 2 above)
Comment (alheinecke) [reply src/generator_common_aarch64.h:21]: nit: i_is_gather is like i_is_store for regular loads and should be in the end of the arg list

File: src/generator_mateltwise_gather_scatter_aarch64.h [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+void libxsmm_generator_gather_scatter_cols_aarch64_microkernel( libxsmm_generated_code*                        io_generated_code,
+    libxsmm_loop_label_tracker*                    io_loop_label_tracker,
+    libxsmm_mateltwise_gp_reg_mapping*             i_gp_reg_mapping,
+    const libxsmm_mateltwise_kernel_config*        i_micro_kernel_config,
+    const libxsmm_meltw_descriptor*                i_mateltwise_desc );
+
+LIBXSMM_API_INTERN
+void libxsmm_generator_gather_scatter_rows_aarch64_mn_loop_unrolled( libxsmm_generated_code*                        io_generated_code,
+    libxsmm_mateltwise_gp_reg_mapping*             i_gp_reg_mapping,
+    unsigned int                                   m_unroll_factor,
```
Comment (alheinecke) [inline src/generator_mateltwise_gather_scatter_aarch64.h:27]: nit: the i_ for input are missing
nit: why not const int?

File: src/generator_mateltwise_gather_scatter_aarch64.h [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+void libxsmm_generator_gather_scatter_rows_aarch64_microkernel( libxsmm_generated_code*                        io_generated_code,
+    libxsmm_loop_label_tracker*                    io_loop_label_tracker,
+    libxsmm_mateltwise_gp_reg_mapping*             i_gp_reg_mapping,
+    const libxsmm_mateltwise_kernel_config*        i_micro_kernel_config,
+    const libxsmm_meltw_descriptor*                i_mateltwise_desc );
+
+LIBXSMM_API_INTERN
+void libxsmm_generator_gather_scatter_offs_aarch64_mn_loop_unrolled( libxsmm_generated_code*                        io_generated_code,
+    libxsmm_mateltwise_gp_reg_mapping*             i_gp_reg_mapping,
+    unsigned int                                   m_unroll_factor,
```
Comment (alheinecke) [inline src/generator_mateltwise_gather_scatter_aarch64.h:55]: nit: the i_ for input are missing
nit: why not const int?

File: src/generator_mateltwise_gather_scatter_aarch64.c [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+#define USE_ENV_TUNING
+#endif
+
+LIBXSMM_API_INTERN
+void libxsmm_generator_gather_scatter_cols_aarch64_microkernel( libxsmm_generated_code*                        io_generated_code,
+    libxsmm_loop_label_tracker*                    io_loop_label_tracker,
+    libxsmm_mateltwise_gp_reg_mapping*             i_gp_reg_mapping,
+    const libxsmm_mateltwise_kernel_config*        i_micro_kernel_config,
+    const libxsmm_meltw_descriptor*                i_mateltwise_desc ) {
+  unsigned int m, n, im = 0, m_remainder_elements, m_trips, max_m_unrolling = 4, m_unroll_factor = 1, m_trips_loop = 0, peeled_m_trips = 0;
```
Comment (alheinecke) [inline src/generator_mateltwise_gather_scatter_aarch64.c:33]: nit: some local variable have correctly l_ as prefix, others don't

File: src/generator_mateltwise_gather_scatter_aarch64.c [hunk 7]
```diff
@@ -0,0 +1,804 @@
...
+    unsigned int                                   vlen,
+    unsigned int                                   m_remainder_elements,
+    unsigned int                                   mask_reg,
+    unsigned int                                   mask_reg_full_frac_vlen,
+    unsigned int                                   is_gather,
+    unsigned int                                   gp_idx_mat_reg,
+    unsigned int                                   gp_reg_mat_reg,
+    unsigned int                                   dtype_size_reg_mat,
+    unsigned int                                   ld_reg_mat ) {
+  unsigned char l_is_sve = (io_generated_code->arch >= LIBXSMM_AARCH64_SVE128) && (io_generated_code->arch <= LIBXSMM_AARCH64_ALLFEAT);
```
Comment (alheinecke) [inline src/generator_mateltwise_gather_scatter_aarch64.c:227]: nit: some local variable have correctly l_ as prefix, others don't

File: src/generator_mateltwise_gather_scatter_aarch64.c [hunk 8]
```diff
@@ -0,0 +1,804 @@
...
+                                               (long long)l_ld_bytes_reg*n_unroll_factor );
+}
+
+LIBXSMM_API_INTERN
+void libxsmm_generator_gather_scatter_rows_aarch64_microkernel( libxsmm_generated_code*                        io_generated_code,
+    libxsmm_loop_label_tracker*                    io_loop_label_tracker,
+    libxsmm_mateltwise_gp_reg_mapping*             i_gp_reg_mapping,
+    const libxsmm_mateltwise_kernel_config*        i_micro_kernel_config,
+    const libxsmm_meltw_descriptor*                i_mateltwise_desc ) {
+  unsigned int m, n, im = 0, m_remainder_elements, m_trips, max_m_unrolling = 4, m_unroll_factor = 1, n_unroll_factor = 4, m_trips_loop = 0, peeled_m_trips = 0;
```
Comment (alheinecke) [inline src/generator_mateltwise_gather_scatter_aarch64.c:304]: nit: some local variable have correctly l_ as prefix, others don't

File: src/generator_mateltwise_gather_scatter_aarch64.c [hunk 9]
```diff
@@ -0,0 +1,804 @@
...
+    unsigned int                                   vlen,
+    unsigned int                                   m_remainder_elements,
+    unsigned int                                   mask_reg,
+    unsigned int                                   mask_reg_full_frac_vlen,
+    unsigned int                                   is_gather,
+    unsigned int                                   gp_idx_mat_reg,
+    unsigned int                                   gp_reg_mat_reg,
+    unsigned int                                   dtype_size_reg_mat,
+    unsigned int                                   ld_reg_mat ) {
+  unsigned char l_is_sve = (io_generated_code->arch >= LIBXSMM_AARCH64_SVE128) && (io_generated_code->arch <= LIBXSMM_AARCH64_ALLFEAT);
```
Comment (alheinecke) [inline src/generator_mateltwise_gather_scatter_aarch64.c:525]: nit: some local variable have correctly l_ as prefix, others don't

File: src/generator_mateltwise_gather_scatter_aarch64.c [hunk 10]
```diff
@@ -0,0 +1,804 @@
...
+                                               (long long)l_ld_bytes_reg*n_unroll_factor );
+}
+
+LIBXSMM_API_INTERN
+void libxsmm_generator_gather_scatter_offs_aarch64_microkernel( libxsmm_generated_code*                        io_generated_code,
+    libxsmm_loop_label_tracker*                    io_loop_label_tracker,
+    libxsmm_mateltwise_gp_reg_mapping*             i_gp_reg_mapping,
+    const libxsmm_mateltwise_kernel_config*        i_micro_kernel_config,
+    const libxsmm_meltw_descriptor*                i_mateltwise_desc ) {
+  unsigned int m, n, m_remainder_elements, m_trips, max_m_unrolling = 4, m_unroll_factor = 1, n_unroll_factor = 4, m_trips_loop = 0, peeled_m_trips = 0;
```
Comment (alheinecke) [inline src/generator_mateltwise_gather_scatter_aarch64.c:611]: nit: some local variable have correctly l_ as prefix, others don't

File: src/generator_mateltwise_reduce_aarch64.c [hunk 11]
```diff
@@ -241,7 +241,8 @@ void libxsmm_generator_reduce_cols_aarch64_microkernel( libxsmm_generated_code*
   libxsmm_aarch64_asimd_tupletype asimd_type = (i_micro_kernel_config->datatype_size_in == 4) ? LIBXSMM_AARCH64_ASIMD_TUPLETYPE_4S : LIBXSMM_AARCH64_ASIMD_TUPLETYPE_2D;
   unsigned char is_sve = (io_generated_code->arch >= LIBXSMM_AARCH64_SVE128) && (io_generated_code->arch <= LIBXSMM_AARCH64_ALLFEAT);
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_aarch64.c:242]: nit: some local variable have correctly l_ as prefix, others don't

**Review by alheinecke — COMMENTED** [review]
mostly nits for code quality on the new files.
also: there is a generator_mateltwise_aarch64.c and generator_mateltwise_aarch64_sve.c file (which existed before the PR) but there seems to be duplication. Can we just merge these files as in the x86 case? I guess this an earlier oversight, but this PR makes duplication worse, so let's clean up.

**Comment by egeor** [comment]
This PR should be good to merge now.

File: samples/xgemm/kernel_test/generate_gemm_test_scripts.sh [hunk 12]
```diff
@@ -136,8 +136,8 @@ for BINARY_POSTOP in 0 1; do
 
                 if [ "$CVNNI" == '1' ] ; then
                   if [ "$PREC" == 'BF16' ] ; then
-                    NSTART=2
-                    NSTEP=2
+                    NSTART=4
```
Comment (alheinecke) [inline samples/xgemm/kernel_test/generate_gemm_test_scripts.sh:139]: should that be for MMLA only? CPX should be tested for NSTEP=2

File: src/generator_mateltwise.c [hunk 13]
```diff
@@ -13,18 +13,16 @@
...
-#include "generator_mateltwise_aarch64_sve.h"
 
 LIBXSMM_API
 void libxsmm_generator_mateltwise_kernel( libxsmm_generated_code*          io_generated_code,
                                           const libxsmm_meltw_descriptor*  i_mateltw_desc ) {
   /* generate kernel */
   if ( (io_generated_code->arch >= LIBXSMM_X86_GENERIC) && (io_generated_code->arch <= LIBXSMM_X86_ALLFEAT) ) {
     libxsmm_generator_mateltwise_sse_avx_avx512_kernel( io_generated_code, i_mateltw_desc );
-  } else if ( (io_generated_code->arch >= LIBXSMM_AARCH64_V81) && (io_generated_code->arch <= LIBXSMM_AARCH64_V82) ) {
+  } else if ( ((io_generated_code->arch >= LIBXSMM_AARCH64_V81) && (io_generated_code->arch <= LIBXSMM_AARCH64_V82)) ||
```
Comment (alheinecke) [inline src/generator_mateltwise.c:23]: Should that if incl. all ARM archs?

**Review by alheinecke — COMMENTED** [review]
2 new comments

File: samples/xgemm/kernel_test/generate_gemm_test_scripts.sh (hunk 12 above)
Comment (alheinecke) [reply samples/xgemm/kernel_test/generate_gemm_test_scripts.sh:139]: In Fact there is a patch for mmla files in the end. So regular bf16 stuff should be allow-listed

File: src/generator_mateltwise.c (hunk 13 above)
Comment (egeor) [reply src/generator_mateltwise.c:23]: Correct, the if branches are identical now.

File: samples/xgemm/kernel_test/generate_gemm_test_scripts.sh (hunk 12 above)
Comment (egeor) [reply samples/xgemm/kernel_test/generate_gemm_test_scripts.sh:139]: Yeah, was not sure how to deal with this -- allow-listing this and using the mmla tests is the right thing

File: src/generator_mateltwise.c (hunk 13 above)
Comment (alheinecke) [reply src/generator_mateltwise.c:23]: So what about APPL_M1. It should work because it's 82, but why error we out?

Comment (egeor) [reply src/generator_mateltwise.c:23]: That's what iImeant in my earlier comment, all aarch64 shoould go there (incl M1)

File: samples/xgemm/kernel_test/generate_gemm_test_scripts.sh (hunk 12 above)
Comment (alheinecke) [reply samples/xgemm/kernel_test/generate_gemm_test_scripts.sh:139]: ah OK, then I wait for your commit and merge

Comment (egeor) [reply samples/xgemm/kernel_test/generate_gemm_test_scripts.sh:139]: But do we want also do unit test the bfdot kernels? then these would not be allow-listed but we would have to set an env var or so

Comment (alheinecke) [reply samples/xgemm/kernel_test/generate_gemm_test_scripts.sh:139]: Right now we would need to add unit test for BFDOT. 
@hfp: Can we add a second GEMM test for Gravition exposing "LIBXSMM_AARCH64_USE_BFDOT=1" and in the case the would use a different allow list, e.g. the test tool would check the env variable and then loads the a different allow list file. Is this possible?

**Review by alheinecke — APPROVED** [review]

**Pull request merged**